contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+
+### Non Goals
+Everything covered in the [Monitoring Architecture](https://github.com/kubernetes/kubernetes/blob/master/docs/design/monitoring_architecture.md) design doc.  This includes the third party metrics pipeline, and the methods by which the metrics found in this proposal are provided to other kubernetes components.
+
+Integration with CRI will not be covered in this proposal.  In future proposals, integrating with CRI may provide a better abstraction of information required by the core metrics pipeline to collect metrics.
+
+## Design
+
+This design covers only the internal Core Metrics Pipeline.
+
+High level requirements for the design are as follows:
+ - Do not break existing users.  We should continue to provide the full summary API by default.
+ - The kubelet collects the minimum possible number of metrics for full kubernetes functionality.
+ - Code for collecting core metrics resides in the kubernetes codebase.
+ - Metrics can be fetched "On Demand", giving the kubelet more up-to-date stats.
+
+Metrics requirements, based on kubernetes component needs, are as follows:
+ - Kubelet
+  - Node-level capacity and availability metrics for Disk and Memory
+  - Pod-level usage metrics for Disk and Memory
+ - Scheduler
+  - Node-level capacity and availability metrics for Disk, CPU, and Memory
+  - Pod-level usage metrics for Disk, CPU, and Memory
+  - Container-level usage metrics for Disk, CPU, and Memory
+ - Horizontal-Pod-Autoscaler (HPA)
+  - Node-level capacity and availability metrics for CPU and Memory
+  - Pod-level usage metrics for CPU and Memory
+
+More details on how I intend to achieve these high level goals can be found in the Implementation Plan.
+
+In order to continue to provide the full summary API, either the kubelet or a stand-alone version of cAdvisor will need to publish these metrics.
+
+This Core Metrics API will be versioned to account for version-skew between kubernetes components.
+
+This proposal purposefully omits many metrics that may eventually become core metrics.  This is by design.  Once metrics are needed for an internal use case, they can be added to the core metrics API.
+
+### Proposed Core Metrics API:
+
+An important difference between the current summary api and the proposed core metrics api is that per-pod stats in the core metrics api contain only usage data, and not capacity-related statistics.  This is more accurate since a pod's resource capacity is really defined by its "requests" and "limits", and it is a better reflection of how the kubelet uses the data.  The kubelet finds which resources are constrained using node-level capacity and availability data, and then chooses which pods to take action on based on the pod's usage of the constrained resource.  If neccessary, capacity for resources a pod consumes can still be correlated with node-level resources using this format of stats.
```

Comment:
Sorry, maybe I am being too philosophical.  All I am trying to convey is that it doesnt make sense for pods to have a "capacity".  Rather, they have usage (xxx bytes of disk, for example), and a resource they are using (filesystem device, for example).  A "Node" has resources, and a "Pod" has usage.  But I can still find how much of a resource is available by correlating the two.